src/main/scala/com/salesforce/spark/carrot/RabbitUtils.scala

```diff
@@ -0,0 +1,206 @@
+  (
+    ssc: StreamingContext,
+    rmqConfig: RMQConfig,
+    storageLevel: StorageLevel = StorageLevel.MEMORY_AND_DISK_SER_2
+    ): ReceiverInputDStream[T] = new RMQInputDStream[T, U](ssc, rmqConfig, storageLevel)
+}
+
+/**
+ * Create an input stream that pulls messages from Rabbit MQ.
+ * @param ssc_         StreamingContext object
+ * @param rmqConfig   RMQConfig object containing RabbitMQ specific configurations
+ * @param storageLevel Storage level to use for storing the received objects
+ */
+class RMQInputDStream[T: ClassTag, U <: Decoder[_] : ClassTag]
+(
+  @transient ssc_ : StreamingContext,
+  rmqConfig: RMQConfig,
+  storageLevel: StorageLevel)
+  extends ReceiverInputDStream[T](ssc_) {
+
+  def getReceiver(): Receiver[T] = new RMQReceiver[T, U](rmqConfig, storageLevel)
+
+}
+
+class ExponentialDelay(initialDelay: Long = 5, maxDelay: Long = 1000) {
+  var waitTime = initialDelay
+
+  def reset() = waitTime = initialDelay
+
+  def pulse() = {
+    Thread.sleep(waitTime)
+    waitTime *= 2
+    waitTime = math.min(waitTime, maxDelay)
+  }
+}
+
+/**
+ * Connection factory that automatically sets configs from RMQConfig and has an exponential delay and alternative host mechanic.
+ *
+ * See ConnectionFactory for details on RabbitMQ specific parameters.
+ *
+ * @param rmqConfig RMQConfig that defines RMQ behavior
+ */
+class ConnectionFactoryWithFailover(rmqConfig: RMQConfig) extends ConnectionFactory() {
+  var ifPrimaryHost = true
+  this.setHost(rmqConfig.hostname)
+  this.setPort(rmqConfig.port)
+  rmqConfig.username.foreach(this.setUsername _)
+  rmqConfig.password.foreach(this.setPassword _)
+
+  def toggleHost = {
+    if (ifPrimaryHost) this.setHost(rmqConfig.alternativeHost)
+    else if (!ifPrimaryHost) this.setHost(rmqConfig.hostname)
+
+    ifPrimaryHost = !ifPrimaryHost
+  }
+
+  /**
+   * This is an exponential delay on the calls to RabbitMQ.  If a machine is unresponsive, exponentially wait.
+   * Also swap between this host and the alternative host if there is one.
+   */
+  override def newConnection: Connection = {
+    val connectionExponentialDelay = new ExponentialDelay(1000, 600000)
+    while (true) {
+      try {
+        val connection = super.newConnection
+        connectionExponentialDelay.reset()
+        return connection
+      } catch {
+        case ioe: ConnectException => {
+          connectionExponentialDelay.pulse()
+          toggleHost
+        }
+        case e: Exception => throw e
+      }
+    }
+    throw new IllegalStateException("This code path should be inaccessible.")
+  }
+}
+
+class RMQReceiver[T: ClassTag, U <: Decoder[_] : ClassTag]
+(
+  rmqConfig: RMQConfig,
+  @transient storageLevel: StorageLevel
+  )
+  extends Receiver[T](storageLevel) with Logging with Serializable {
+
+  def onStart {
+    // Start the thread that receives data over a connection
+    new Thread("Socket Receiver") {
+      override def run {
+        receive
```

Comment:
@suryasev you code is similar to this one: https://github.com/apache/spark/blob/d83c2f9f0b08d6d5d369d9fae04cdb15448e7f0d/streaming/src/main/scala/org/apache/spark/streaming/dstream/SocketInputDStream.scala#L48

As I mentioned above, notice some important differences:
1. the `def receive()` function handles the errors properly, but not in your case
2. it might be a good idea to set `setDaemon(true)` on Thread start